Mode handoff flow

digitarald opened the pull request:
# Mode Handoffs Implementation

Adds a `handoffs` metadata field to chat modes that provides suggested follow-up prompts in a collapsible UI widget, enabling guided workflows for iterative development tasks.

## Demo

<img width="1466" height="891" alt="image" src="https://github.com/user-attachments/assets/142080a8-edc3-4594-b2ea-55b991b60133" />


**Example chat mode file:**
```yaml
---
description: Research and draft an implementation plan
tools: ['search', 'executePrompt', 'digitarald.agent-memory/memory', 'todos']
handoffs: 
  - label: Start Implementation
    agent: agent
    prompt: Start implementation
  - label: Move to Editor
    agent: agent
    prompt: Save the plan
    send: true
---
Your goal is to draft a clear, detailed, and actionable plan that addresses the user's request.
```

---

## Implementation Summary

### Backend (✅ Complete)
- **Data Model**: Added `handoffs?: Record<string, string>` to chat mode interfaces
- **Observable State**: Integrated with existing observable patterns for reactive updates
- **Caching**: Full serialization/deserialization support
- **Validation**: Added `handoffs` to valid attributes (19/19 tests passing)

### Frontend (✅ Complete)
- **Widget Component** (`chatSuggestNextWidget.ts`): Reuses existing `chat-welcome-view-suggested-prompts` CSS classes for visual consistency
- **Title Format**: "Proceed from {Mode}" using `IChatMode.name` property
- **Integration**: Widget appears above chat input when response starts in mode with handoffs
- **User Interaction**: Clicking prompt inserts `/{promptId} {description}` into input
- **Layout**: Natural flow positioning (not sticky at top)

---

## Data Flow

```
User sends message → Response starts → renderChatSuggestNextWidget() called
→ currentModeObs.get() retrieves IChatMode with handoffs
→ Widget renders with "Proceed from {Mode}" title and prompt buttons
→ User clicks prompt → Inserts into input → Widget hides
```

---

## Key Technical Decisions

### 1. Mode Detection Fix
**Problem**: `currentModeKind` returns only enum values, making custom modes indistinguishable  
**Solution**: Use `currentModeObs.get()` to access actual `IChatMode` object with unique ID, name, and handoffs

### 2. CSS Reuse Strategy
**Approach**: Reuse existing `chat-welcome-view-suggested-prompts` classes from welcome view  
**Benefits**: Visual consistency, reduced code duplication, proven accessibility patterns  
**Custom overrides**: Only positioning and layout adjustments in `chat.css`

### 3. Widget Positioning
**Approach**: Natural flow above input (not sticky at top)  
**Rationale**: Better UX, appears where users expect it, simpler layout logic

### 4. Height Event Timing (Critical Fix)
**Issue**: Using `queueMicrotask()` caused stale height reads and text overlay  
**Solution**: Fire height change events **immediately** after DOM updates  
**Learning**: Input-adjacent widgets differ from tree-rendered content parts

---

## Outstanding Issues from Review

### 🟢 Resolved Issues
- ✅ Quick Chat limitation (widget properly scoped to Quick Chat context only)
- ✅ Widget in empty chat (fixed with debouncing via `RunOnceScheduler`)
- ✅ Widget shows when clearing chat (added `cancel()` call in `clear()` method)
- ✅ Widget visible on empty chat load (added content and quick chat checks)
- ✅ Widget persists during mode change (added `hide()` before conditional show)

---

## Files Modified

**Backend (5 files)**
- `chatModes.ts` - Core data model
- `promptsService.ts` - Interface definitions
- `promptsServiceImpl.ts` - Implementation
- `newPromptsParser.ts` - YAML parsing
- `promptValidator.ts` - Validation rules

**Frontend (3 files)**
- `chatSuggestNextWidget.ts` (NEW) - Widget component
- `chatWidget.ts` - Integration and lifecycle
- `chat.css` - Styling overrides

**Tests (1 file)**
- `promptValidator.test.ts` - Validation tests (19/19 passing)

---

## Testing Status

### ✅ Manual Testing (Complete)
- Widget visibility lifecycle verified
- Keyboard navigation works
- Mode switching updates correctly
- Visual consistency confirmed
- Theme compatibility checked

### ⚠️ Automated Testing (Missing)
- No unit tests for `ChatSuggestNextWidget`
- No integration tests
- Only validator tests exist (passing)

---

## Visual Design

**Styling Approach**: Reuses `chat-welcome-view-suggested-prompts` classes
- Container: `.chat-welcome-view-suggested-prompts` with flexbox layout
- Title: `.chat-welcome-view-suggested-prompts-title` with uppercase styling
- Buttons: `.chat-welcome-view-suggested-prompt` with border and hover states
- Text: `.chat-welcome-view-suggested-prompt-title` for prompt labels

**Custom Overrides** (in `chat.css`):
- Positioning: `position: relative` instead of absolute
- Padding: `32px 16px 8px 16px` for title and button spacing
- Alignment: `justify-content: flex-start` to left-align buttons

---

## Success Metrics

**Quantitative**:
- ✅ Zero TypeScript compilation errors
- ✅ All validator tests passing (19/19)

**Qualitative**:
- ✅ Visual consistency with existing welcome view prompts
- ✅ Smooth layout updates (no jank)
- ✅ Works across all themes
- ✅ Intuitive keyboard navigation

---

## Next Steps

1. **[MEDIUM]** Add unit tests for widget component
2. **[LOW]** Add integration tests for chat widget interaction
3. **[PENDING]** Wait for validator review from @aeschli

---

## Compilation Status

✅ All TypeScript compiles successfully
- Core - Build: Running (no errors)
- Ext - Build: Running (no errors)

digitarald committed:
Mode handoff flow

digitarald committed:
Show/hide widget on mode switching

digitarald committed:
Fix tests

OrenMe commented:
Could we also have the entry reference a *.prompt.md file, it already has description and agent fields and the actual prompt

bhavyaus commented on src/vs/workbench/contrib/chat/common/promptSyntax/languageProviders/promptValidator.ts:
```diff
@@ -307,7 +307,7 @@ export class PromptValidator {
 const validAttributeNames = {
 	[PromptsType.prompt]: ['description', 'model', 'tools', 'mode'],
 	[PromptsType.instructions]: ['description', 'applyTo', 'excludeAgent'],
-	[PromptsType.mode]: ['description', 'model', 'tools', 'advancedOptions']
```
cc: @aeschli  for prompt validator changes here.

bhavyaus commented on src/vs/workbench/contrib/chat/browser/chatContentParts/chatSuggestNextWidget.ts:
```diff
@@ -0,0 +1,216 @@
+	 * Updates the ARIA label for the widget to announce expanded/collapsed state
+	 */
+	private updateAriaLabel(element: HTMLElement, modeLabel: string, expanded: boolean): void {
+		const label = localize('chat.suggestNext.title', "Continue with {0}?", modeLabel);
```
This label feels a bit confusing. Do we even need it?
If I wanted to act on one of the items in the list, I’d just click it. Ignoring those options already implies I’m continuing in the current mode.
“Continue with RubberDuck” reads like I need to click it just to keep chatting, which isn’t intuitive.

<img width="869" height="151" alt="Image" src="https://github.com/user-attachments/assets/ce4cd55c-59ef-4326-a2f0-490e17d1ca01" />

bhavyaus replied:
cc: @unthinkmedia

digitarald replied:
Fixed with latest UI :"Procseed with {mode}"

bhavyaus commented on src/vs/workbench/contrib/chat/browser/chatWidget.ts:
```diff
@@ -2240,6 +2295,7 @@ export class ChatWidget extends Disposable implements IChatWidget {
 
 		this.renderer.updateViewModel(this.viewModel);
 		this.updateChatInputContext();
+		this.renderChatSuggestNextWidget();
```
If yes, we should consider debouncing this with `RunOnceScheduler` to avoid performing this render too many times

digitarald replied:
Should be fixed

digitarald committed:
Address PR feedback: remove subtitle, exclude quick chat, add debouncing for suggest next widget

digitarald committed:
Show next steps widget only after request completion, not on message send

digitarald committed:
Reset promptSyntax changes that landed in main

The handOffs property logic has landed in main branch via another PR.
Resetting these files to avoid conflicts and use the upstream implementation.

digitarald committed:
Fixed handoff props

digitarald committed:
Improve chat suggest next widget layout and styling

digitarald committed:
Fix mode handoff widget visibility on clear and mode changes

digitarald committed:
Prevent suggest widget re-show on clear

bhavyaus committed:
reset styling and clean up chatWidget

bhavyaus committed:
Schedule welcome view render instead of invoking it directly

Use the existing _welcomeRenderScheduler.schedule() when the view model has no items
so welcome view rendering is debounced and avoids potential reentrant/cyclic calls.

bhavyaus committed:
update css

bhavyaus commented on src/vs/workbench/contrib/chat/browser/chatWidget.ts:
```diff
@@ -2136,6 +2209,8 @@ export class ChatWidget extends Disposable implements IChatWidget {
 			this.refreshParsedInput();
 			this.renderFollowups();
+			this.chatSuggestNextWidget.hide();
+			this._chatSuggestNextScheduler.schedule();
```
nit fix this.

bhavyaus committed:
Schedule suggest-next widget render instead of hiding it synchronously

bhavyaus committed:
Use single quotes for YAML lines in "mode with handoffs attribute" test

bhavyaus committed:
Update tests

digitarald committed:
Account for suggest-next widget height

bhavyaus commented on src/vs/workbench/contrib/chat/browser/chatWidget.ts:
```diff
@@ -2644,6 +2729,7 @@ export class ChatWidget extends Disposable implements IChatWidget {
 			welcomeOffset = Math.max(welcomeOffset - this.input.editSessionWidgetHeight, 0);
 		}
 		welcomeOffset = Math.max(welcomeOffset - this.input.attachmentsHeight, 0);
+		welcomeOffset = Math.max(welcomeOffset - chatSuggestNextWidgetHeight, 0);
```
I don't think this is needed since suggest widget is never rendered when welcome is shown

bhavyaus committed:
Nit: remove height update for welcome content

bpasero commented:
@bhavyaus @digitarald what is going on with the schedulers to render UI 🤯 

https://github.com/microsoft/vscode/blob/afd4a5ee0cda0902c789df5deff9effd7736545f/src/vs/workbench/contrib/chat/browser/chatWidget.ts#L551-L560

That can not be our programming model that UI things are being scheduled at random intervals. Why are we applying that pattern now again for suggested prompts?
